sdk/iothub/src/service/requests/create_or_update_device_identity_builder.rs

```diff
@@ -0,0 +1,183 @@
+use std::convert::TryInto;
+
+use azure_core::errors::AzureError;
+use http::Method;
+use serde::Serialize;
+
+use crate::service::resources::{
+    identity::DesiredCapability, identity::IdentityOperation, AuthenticationMechanism,
+    AuthenticationType, DeviceCapabilities, Status, SymmetricKey, X509ThumbPrint,
+};
+use crate::service::responses::DeviceIdentityResponse;
+use crate::service::{IoTHubError, ServiceClient, API_VERSION};
+
+/// The CreateOrUpdateDeviceIdentityBuilder is used to construct a new device identity
+/// or the update an existing one.
+pub struct CreateOrUpdateDeviceIdentityBuilder<'a> {
+    service_client: &'a ServiceClient,
+    authentication: Option<AuthenticationMechanism>,
+    capabilities: DeviceCapabilities,
+    device_id: Option<String>,
+    status: Option<Status>,
+    etag: Option<String>,
+    operation: IdentityOperation,
+}
+
+impl<'a> CreateOrUpdateDeviceIdentityBuilder<'a> {
+    pub(crate) fn new(
+        service_client: &'a ServiceClient,
+        operation: IdentityOperation,
+        etag: Option<String>,
+    ) -> Self {
+        Self {
+            service_client,
+            authentication: None,
+            capabilities: DeviceCapabilities::default(),
+            device_id: None,
+            status: None,
+            etag,
+            operation,
+        }
+    }
+
+    /// Set authentication to SAS on the device
+    pub fn authentication_using_sas<S, T>(mut self, primary_key: S, secondary_key: T) -> Self
```

Comment:
Should we use [the `setters!` macro](https://github.com/Azure/azure-sdk-for-rust/blob/02949f433aeb2ac58ada7063851cb3921b82e4f9/sdk/cosmos/src/requests/create_collection_builder.rs#L35-L40) for these setters?

Reply:
Based on your other comment I moved the required fields over to `execute`. There is one optional field but I don't believe I can use the `setters!` macro for this.